File: Laravel/app/Services/MyBackupService.php
<?php
/**
 * Created by PhpStorm.
 * User: tsao
 * Date: 2017/10/24
 * Time: 下午 07:27
 */

namespace App\Services;


use App\Services\Configs\ConfigManager;
use App\Services\Configs\JsonManager;
use App\Services\Configs\ScheduleManager;
use App\Services\Tasks\TaskDispatcher;

/**
 * 備份執行類別
 * Class MyBackupService
 * @author steve.tsao
 * @package App\Services
 */
class MyBackupService
{
    /**
     * @var array 存放JsonManager物件陣列
     */
    private $managers = [];

    /**
     * @var TaskDispatcher $taskDispatcher
     */
    private $taskDispatcher;

    /**
     * 建構子
     * MyBackupService constructor.
     * @author steve.tsao
     */
    public function __construct()
    {
        $this->managers = [
            new ConfigManager(),
            new ScheduleManager(),
        ];

        $this->taskDispatcher = new TaskDispatcher();

        $this->Init();
    }

    /**
     * 物件初始化
     * @author steve.tsao
     */
    private function Init()
    {
        $this->ProcessJsonConfigs();
    }

    /**
     * 處理JSON檔案執行設定
     * @author steve.tsao
     */
    private function ProcessJsonConfigs()
    {
        // 呼叫JsonManager子類別實作的ProcessJsonConfig函數
        collect($this->managers)->each(function (JsonManager $item) {
            $item->ProcessJsonConfig();
        });
    }

    /**
     * 簡單備份
     * @author steve.tsao
     */
    public function SimpleTask()
    {
        $this->taskDispatcher->SimpleTask($this->managers);
    }

    /**
     * 排程備份
     * @author steve.tsao
     */
    public function ScheduleTask()
    {
        $this->taskDispatcher->ScheduleTask($this->managers);
    }
}

File: NetCore/ConsoleOop/ConsoleOop/MyBackupService.cs
using System.Collections.Generic;
using ConsoleOop.Configs;
using ConsoleOop.Tasks;

namespace ConsoleOop
{
    /// <summary>
    /// 備份類別
    /// </summary>
    class MyBackupService
    {
        /// <summary>
        /// 存放JsonManager物件陣列
        /// </summary>
        private List<JsonManager> managers = new List<JsonManager>();

        /// <summary>
        /// 備份執行方式
        /// </summary>
        private TaskDispatcher taskDispatcher;

        /// <summary>
        /// 建構子 依賴注入
        /// </summary>
        public MyBackupService()
        {
            this.managers.Add(new ConfigManager());
            this.managers.Add(new ScheduleManager());
            this.taskDispatcher = new TaskDispatcher();

            this.Init();
        }

        /// <summary>
        /// 物件初始化
        /// </summary>
        private void Init()
        {
            this.ProcessJsonConfigs();
        }

        /// <summary>
        /// 簡單備份
        /// </summary>
        public void SimpleTask()
        {
            this.taskDispatcher.SimpleTask(this.managers);
        }

        /// <summary>
        /// 排程備份
        /// </summary>
        public void ScheduledTask()
        {
            this.taskDispatcher.ScheduledTask(this.managers);
        }

        /// <summary>
        /// 處理JSON檔案
        /// </summary>
        private void ProcessJsonConfigs()
        {
            this.managers.ForEach(delegate (JsonManager jsonManager)
            {
                jsonManager.ProcessJsonConfig();
            });
        }
    }
}
